### PR TITLE
Port tests in `csv_files.rs` to sqllogictest

### LINKED ISSUE #8197: Port tests in `csv_files.rs` to sqllogictest
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/arrow-datafusion/issues/6195. We are trying to
1. keep the DataFusion code base clean and easy to contribute to and modify
2. Understand what features are covered and what are not

Part of doing so is having a single location for most test coverage so new tests can be added easily and people can follow the existing patterns easily

### Describe the solution you'd like

Port the tests in https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sql/csv_files.rs to
`csv_files.slt` in https://github.com/apache/arrow-datafusion/tree/main/datafusion/sqllogictest/test_files

Note you can create catalogs and schema via SQL now

Catalogs: https://arrow.apache.org/datafusion/user-guide/sql/ddl.html#create-database
Schema: https://arrow.apache.org/datafusion/user-guide/sql/ddl.html#create-schema

Notes:
1. Instructions for running sqllogitest are here: https://github.com/apache/arrow-datafusion/tree/main/datafusion/sqllogictest

### Describe alternatives you've considered

_No response_

### Additional context

I think these are good first issues as they teach the contributor about the DataFusion codebase and tests, as well as see the end user apis in practice.

### PR DESCRIPTION
**Pull request by Asura7969**
## Which issue does this PR close?

Closes #8197 .

## What changes are included in this PR?

2 parts：
* Add `quote` & `escape` for `create external table` in sql（The api method is already supported）
* Port tests in csv_files.rs to sqllogictest
## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/core/src/datasource/listing_table_factory.rs [hunk 1]
```diff
@@ -73,6 +73,8 @@ impl TableProviderFactory for ListingTableFactory {
                 CsvFormat::default()
                     .with_has_header(cmd.has_header)
                     .with_delimiter(cmd.delimiter as u8)
+                    .with_quote(cmd.options.get("quote").map_or(b'"', |x| x.as_bytes()[0]))
+                    .with_escape(cmd.options.get("escape").map_or(None, |x| Some(x.as_bytes()[0])))
                     .with_file_compression_type(file_compression_type),
```
Comment (Asura7969) [inline datafusion/core/src/datasource/listing_table_factory.rs:78]: I'm not sure if this is the best way, maybe we can open another issue to improve it

File: datafusion/sqllogictest/test_files/csv_files.slt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+statement ok
+CREATE EXTERNAL TABLE csv_custom_quote (
+    c1  VARCHAR DEFAULT NULL,
+    c2  VARCHAR DEFAULT NULL
+)
+STORED AS CSV
+WITH HEADER ROW
+DELIMITER ','
+OPTIONS ('quote' '~')
+LOCATION '../../testing/data/csv/csv_custom_quote.csv';
```
Comment (Asura7969) [inline datafusion/sqllogictest/test_files/csv_files.slt:31]: This file is needed here, but I don't know how to create it.

The contents of the file(`csv_custom_quote.csv`) are as follows:

c1,c2
\~id0~,\~value0~
\~id1~,\~value1~
\~id2~,\~value2~
\~id3~,\~value3~
\~id4~,\~value4~
\~id5~,\~value5~
\~id6~,\~value6~
\~id7~,\~value7~
\~id8~,\~value8~
\~id9~,\~value9~

File: datafusion/sqllogictest/test_files/csv_files.slt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+statement ok
+CREATE EXTERNAL TABLE csv_custom_escape (
+    c1  VARCHAR DEFAULT NULL,
+    c2  VARCHAR DEFAULT NULL
+)
+STORED AS CSV
+WITH HEADER ROW
+DELIMITER ','
+OPTIONS ('escape' '\')
+LOCATION '../../testing/data/csv/csv_custom_escape.csv';
```
Comment (Asura7969) [inline datafusion/sqllogictest/test_files/csv_files.slt:42]: This is the same

The contents of the file(`csv_custom_escape.csv`):


c1,c2
"id0","value\\"0"
"id1","value\\"1"
"id2","value\\"2"
"id3","value\\"3"
"id4","value\\"4"
"id5","value\\"5"
"id6","value\\"6"
"id7","value\\"7"
"id8","value\\"8"
"id9","value\\"9"

**Comment by alamb** [comment]
Thank you @Asura7969 . This PR appears to be failing CI, so marking it as draft until that is fixed.

**Comment by Asura7969** [comment]
> Thank you @Asura7969 . This PR appears to be failing CI, so marking it as draft until that is fixed.

The reason for this failure is that the test file is missing. Can you help me check [test](https://github.com/apache/arrow-testing/pull/97)?

**Comment by korowa** [comment]
> The reason for this failure is that the test file is missing. Can you help me check https://github.com/apache/arrow-testing/pull/97?

I believe another way is to generate custom-delimited .csv while configuring test context -- reference example might be [csv generation](https://github.com/apache/arrow-datafusion/blob/f310db31b801553bae16ac6a3ef9dc76de7b016a/datafusion/sqllogictest/src/test_context.rs#L104) for `joins.slt`

**Comment by Asura7969** [comment]
> I believe another way is to generate custom-delimited (or escaped) .csv while configuring test context -- reference example might be [csv generation](https://github.com/apache/arrow-datafusion/blob/f310db31b801553bae16ac6a3ef9dc76de7b016a/datafusion/sqllogictest/src/test_context.rs#L104) for `joins.slt`

Thanks for your tip, I'll improve it

**Comment by Asura7969** [comment]
> I believe another way is to generate custom-delimited (or escaped) .csv while configuring test context -- reference example might be [csv generation](https://github.com/apache/arrow-datafusion/blob/f310db31b801553bae16ac6a3ef9dc76de7b016a/datafusion/sqllogictest/src/test_context.rs#L104) for `joins.slt`

I just looked at the relevant code and it doesn't meet the testing requirements, for example(**ddl**):
```sql
CREATE EXTERNAL TABLE csv_custom_quote (
    c1  VARCHAR DEFAULT NULL,
    c2  VARCHAR DEFAULT NULL
)
STORED AS CSV
WITH HEADER ROW
DELIMITER ','
OPTIONS ('quote' '~')  -- this
LOCATION '../../testing/data/csv/csv_custom_quote.csv';
```
> Added quote configuration in sql

Although the api has relevant configuration, it is not supported in sql.
If you have better suggestions, I will improve it.

**Comment by Asura7969** [comment]
> I believe another way is to generate custom-delimited (or escaped) .csv while configuring test context -- reference example might be [csv generation](https://github.com/apache/arrow-datafusion/blob/f310db31b801553bae16ac6a3ef9dc76de7b016a/datafusion/sqllogictest/src/test_context.rs#L104) for `joins.slt`

This PR also involves two attribute（`escape` and `quote`） configurations of the csv file,when creating a table using sql，For example：
```sql
CREATE EXTERNAL TABLE custom_escape (
    c1  VARCHAR DEFAULT NULL,
    c2  VARCHAR DEFAULT NULL
)
STORED AS CSV
WITH HEADER ROW
DELIMITER ','
OPTIONS ('escape' '\')  -- this
LOCATION '../core/tests/data/custom_escape.csv';
```
Using the API to create similar tables supports `escape` and `quote`, but SQL does not support it. Your tip may not cover the SQL method.

Thanks again for the tips.

**Comment by Asura7969** [comment]
I created the test file under the path `datafusion/core/tests/data/`. I don’t know if it is correct.

File: datafusion/core/src/datasource/listing_table_factory.rs (hunk 1 above)
Comment (korowa) [reply datafusion/core/src/datasource/listing_table_factory.rs:78]: TBH I'd prefer to file a separate issued and integrate `QUOTE` & `ESCAPE` into `CreateExternalTable` pipeline. My concern here is that it probably will be confusing for end-users to pass read options into clause reserved for write options (and without any effect on writing CSV).

**Comment by korowa** [comment]
> I created the test file under the path datafusion/core/tests/data/. I don’t know if it is correct.

I suppose, it's more neat to use [enable_testdir](https://github.com/apache/arrow-datafusion/blob/e9b9645ca0da5c6ce3d1e7d8210e442cf565cea0/datafusion/sqllogictest/src/test_context.rs#L116) method of `TextContext` for all temporary test data.

**Comment by Asura7969** [comment]
> I suppose, it's more neat to use [enable_testdir](https://github.com/apache/arrow-datafusion/blob/e9b9645ca0da5c6ce3d1e7d8210e442cf565cea0/datafusion/sqllogictest/src/test_context.rs#L116) method of `TextContext` for all temporary test data.

I used [enable_testdir](https://github.com/apache/arrow-datafusion/blob/e9b9645ca0da5c6ce3d1e7d8210e442cf565cea0/datafusion/sqllogictest/src/test_context.rs#L116) method，thanks for your advice

File: datafusion/sqllogictest/src/test_context.rs [hunk 4]
```diff
@@ -104,6 +104,10 @@ impl TestContext {
                 info!("Registering metadata table tables");
                 register_metadata_tables(test_ctx.session_ctx()).await;
             }
+            "csv_files.slt" => {
```
Comment (alamb) [inline datafusion/sqllogictest/src/test_context.rs:107]: 👍

**Review by alamb — COMMENTED** [review]
Thanks @Asura7969  -- this code looks great to me. Also thank you very much for filing https://github.com/apache/arrow-datafusion/issues/8310

In order to avoid losing test coverage, I think we should fix https://github.com/apache/arrow-datafusion/issues/8310 before merging this one.

Is that something you are able to try @Asura7969 ? Conveniently, you already have written the test for the code in this PR :)

**Comment by Asura7969** [comment]
> Is that something you are able to try @Asura7969 ? Conveniently, you already have written the test for the code in this PR :)

I'll try it this week

**Comment by alamb** [comment]
For anyone following along this is done in https://github.com/apache/arrow-datafusion/pull/8351

**Comment by simicd** [comment]
Thanks for implementing #8351 @Asura7969! Since all tests for #8197 have been ported & with #8885 being merged this draft PR here can be closed?

**Comment by alamb** [comment]
> Thanks for implementing #8351 @Asura7969! Since all tests for #8197 have been ported & with #8885 being merged this draft PR here can be closed?

I agree -- thank you @simicd

**Pull request closed without merging**